designdocs/user-handles.md

```diff
@@ -0,0 +1,323 @@
+
+**Note:** The handle being mapped to MSA ID is the handle with the suffix i.e. full handle.
+
+## Required Extrinsics
+
+### Primitives
+
+``` rust
+
+MsaHandlePayload {
+    handle: &[u8],
+    suffix: u32,
+}
+```
+
+### Create user handle with chosen suffix
+
+ As a network frequency should allow users to choose their own handle and suffix. This extrinsic will allow users to create a handle with a chosen suffix.
+
+``` rust
+Input
+
+* origin - must be a signed origin
+* owner_key - the public key of the user/provider. This is used to verify the signature and resolve the MSA ID.
+* proof - the proof of ownership of the handle, ```MsaHandlePayload``` signed by the user/provider private key
+
+Output
+* Event - `MsaHandleCreated` with the MSA ID and the handle
+
+* Errors - 
+    * `HandleAlreadyExists` if the handle already exists
+    * `InvalidMsaHandle` if the handle is invalid
+    * `Unauthorized` if the signature is invalid
+    * `InvalidSuffix` if the suffix is invalid
+    * `SuffixAlreadyExists` if the suffix already exists
+
+Validation requirements
+* handle must follow handle guidelines.
+* suffix must be in the range of the handle.
+* handle must be unique.
+* handle must be available.
+* suffix must be available.
+* MSA ID must exist. Signing keys must resolve to MSA ID.
+
+Signature requirements
+
+The extrinsic must be signed by the user/provider private key. The signature must be verified on-chain to ensure that the user is the owner of the private key.
+```
+
+### Retire user handle
+
+As a network frequency should allow users to retire their handles. This extrinsic will allow users to retire their handles. Retired handles will be available for reuse after a time period set by governance.
+
+``` rust
+Input
+
+* origin - must be a signed origin
+* owner_key - the public key of the user/provider. This is used to verify the signature and resolve the MSA ID.
+* proof_of_ownership - the proof of ownership of the handle to be retired, ```MsaHandlePayload``` signed by the user/provider private key
+
+Output
+
+* Event - `MsaHandleRetired` with the handle
+
+* Errors - 
+    * `HandleDoesNotExist` if the handle does not exist
+    * `Unauthorized` if the signature is invalid
+
+Validation requirements
+* MSA ID must exist. Signing keys must resolve to MSA ID.
+* handle must exist and be owned by the MSA ID.
+
+Signature requirements
+
+The extrinsic must be signed by the user/provider private key. The signature must be verified on-chain to ensure that the user is the owner of the private key.
+```
+
+**Note:** If retiring a ```Pays::No transaction```? then, we could skip both owner_key and proof_of_ownership.
+
+### Change handle
+
+As a network frequency should allow users to change their handles. This extrinsic will allow users to change their handles. Retired handles will be available for reuse after a time period set by governance.
```

Comment:
How often can a user change their handle?  e.g.  total number of times or elapsed time since last change.  It seems that there should be some limits set by governance.

Reply:
why?

Reply:
Hmm... If someone can continuously change their handle, they could lock up a large number of the common handles.

However suffixes make that a lot harder to do.

It still costs (Capacity) for each new handle or change.